tools/functions_missing_types.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+"""Find the functions in a module missing type annotations.
+
+To use it run
+
+./functions_missing_types.py <module>
+
+and it will print out a list of functions in the module that don't
+have types.
+
+"""
+import argparse
+import ast
+import importlib
+import os
+
+NUMPY_ROOT = os.path.dirname(os.path.join(
+    os.path.abspath(__file__), "..",
+))
+
+# Technically "public" functions (they don't start with an underscore)
+# that we don't want to include.
+EXCLUDE_LIST = {
+    "numpy": {
+        # Stdlib modules in the namespace by accident
+        "absolute_import",
+        "division",
+        "print_function",
+        "warnings",
+        # Accidentally public, deprecated, or shouldn't be used
+        "Tester",
+        "add_docstring",
+        "add_newdoc",
+        "add_newdoc_ufunc",
+        "core",
+        "fastCopyAndTranspose",
+        "get_array_wrap",
+        "int_asbuffer",
+        "oldnumeric",
+        "safe_eval",
+        "set_numeric_ops",
+        "test",
+        # Builtins
+        "bool",
+        "complex",
+        "float",
+        "int",
+        "long",
+        "object",
+        "str",
+        "unicode",
+        # Should use numpy_financial instead
+        "fv",
+        "ipmt",
+        "irr",
+        "mirr",
+        "nper",
+        "npv",
+        "pmt",
+        "ppmt",
+        "pv",
+        "rate",
+        # More standard names should be preferred
+        "alltrue",  # all
+        "sometrue",  # any
+    }
+}
+
+
+class FindAttributes(ast.NodeVisitor):
+    """Find top-level attributes/functions/classes in stubs files.
+
+    Do this by walking the stubs ast. See e.g.
+
+    https://greentreesnakes.readthedocs.io/en/latest/index.html
+
+    for more information on working with Python's ast.
+
+    """
+
+    def __init__(self):
+        self.attributes = set()
+
+    def visit_FunctionDef(self, node):
+        if node.name == "__getattr__":
+            # Not really a module member.
+            return
```

Comment:
Is this actually worth special-casing vs just adding the trivial `__getattr__(attr: str) -> Any` annotation?

Reply:
This is actually filtering out the top-level `__getattr__` annotation in the stubs which is a special PEP 484 thing:

https://www.python.org/dev/peps/pep-0484/#stub-files

to indicate that the stubs are incomplete.